addons/website/static/src/snippets/s_dynamic_snippet/options.js

```diff
@@ -224,13 +227,11 @@ const dynamicSnippetOptions = options.Class.extend({
             if (!this.dynamicFilterTemplates[selectedTemplateId]) {
                 this.$target.get(0).dataset['templateKey'] = dynamicFilterTemplates[0].key;
                 this.isOptionDefault['templateKey'] = true;
-                setTimeout(() => {
-                    this._templateUpdated(dynamicFilterTemplates[0].key, selectedTemplateId);
-                    this._refreshPublicWidgets();
-                });
+                this._templateUpdated(dynamicFilterTemplates[0].key, selectedTemplateId);
+                await this._refreshPublicWidgets();
```

Comment:
What I understand is that it was done here in order to refresh the snippet if its filter changes the templates available for it. E.g. if you drop a debug dynamic snippet, it will start with products. If you switch to a blog filter (New blog posts), after applying the filter to the snippet, it will ask for a rerender of the UI. While rendering the UI it will check if the template is still available (which it isn't since it's a new model) and if not it will assign a new one. Since this is done in `updateUI`, the` _refreshPublicWidget` done by `_select` is passed and done.

I'm assuming it was done there because the code to fetch the new templates was already there?

IMO this isn't done in the best way:
What I would do is
- Fetching the available templates (in a property like `this.templates`) initially in `willStart` (and use them in `renderCustomXML`)
- Selecting a new filter refreshes the templates available for that filter _before_ rerendering the UI
- It also checks if the template is still available and if not applies a new one by changing the data attribute
- the `_select` method automatically refreshes the snippet (dropping the `_refreshPublicWidget` calls)
- `updateUI` re-renders the XML with the correct template values previously fetched

Unfortunately this requires changing some of the methods to async which I don't think I can do on a stable branch. I don't mind doing that on master however with this PR: https://github.com/odoo/odoo/pull/80128

What do you think?

Reply:
Indeed, something like that should be done. But indeed it should be done in master as we don't refactor working things in stable (even if it's technically wrong and potentially breaking in some theoretical cases... it's probably ok). So ok, let's do all that in the PR you mentioned for master (in a dedicated commit).